### DISCUSSION TITLE
RTPS Relay deployment

### DISCUSSION THREAD

**Discussion by smnrgrs** [OpenDDS/OpenDDS#4324] [Q&A] [answered]
We have successfully deployed an instance of the RTPS Relay in ‘the cloud’, where we are also running some DDS participants on the same cloud host. 

This setup allows local participants which have internet access to connect to the relay and successfully pub/sub with the cloud participants.

What if we want an additional cloud host running, perhaps on a different hosting service with a different static IP. Can I use the Relay’s Horizontal scaling to connect between relays on different cloud hosts? I couldn't find the answer to this in the RTPSRelay webinar.

**Comment by jrw972** [comment, marked as answer]
To date, I only have experience with multiple instances in the same network where the trick was to simulate the multicast of SPDP among the different hosts.

If you want to connect two instances that are on completely different networks, then you will need to ensure connectivity for the "Relay Participant" and the horizontal ports.  The main objective is to use fixed ports so that you can configure a firewall appropriately.

For the "Relay Participant", you need to establish connectivity for the SPDP, SEDP, and DATA port.
For SPDP, you can either use the [MulticastRepeater](https://github.com/OpenDDS/cloud-multicast-repeater) or configuration.
If using configuration, then you probably need some combination of `SpdpLocalAddress`, `SpdpSendAddrs`, `UndirectedSpdp`, and `PeriodicDirectedSpdp`.
For SEDP, you probably should disable multicast with `SedpMulticast`, set `SedpLocalAddress`, and maybe `SedpAdvertisedLocalAddress`.
For DATA, you will need to do the same as for SEDP:  disable multicast with `use_multicast`, and set `local_address`, `advertised_address`.

For the horizontal ports, you will need to configure `-HorizontalAddress`.  I think these can take a full address and if they don't, it would be easy to change.

If the Relay Participants and horizontal ports are configured with public IP addresses, then I have concerns about security since 1) the Relay Participant does not use DDS Security and 2) the messages on the horizontal ports are not secure.  Neither of these issues are showstoppers because they have relatively straightforward solutions.  However, it invites us to look at alternatives like connecting the two instances with a VPN for Relay Participant traffic and horizontal traffic.

Instead of thinking about connecting instance to instance, we can think about federating RtpsRelay clusters.
This would be super useful for geographic load balancing, multi-cloud, and hybrid-cloud deployments.
In my mind, the goals would be to 1) make deployment and configuration a lot easier by not requiring advanced configuration, the multicast repeater, VPNs, etc., 2) minimize the cluster-to-cluster messaging, and 3) make discovery more efficient.
